internal/zinc-persist/src/main/scala/sbt/internal/inc/consistent/ConsistentAnalysisFormat.scala

```diff
@@ -0,0 +1,797 @@
+      },
+      2
+    )(writeStamp2(out, _))
+  }
+
+  private[this] def readStamps(in: Deserializer): Stamps = {
+    import VirtualFileUtil._
+    val products =
+      in.readColl[(VirtualFileRef, Stamp), TreeMap[VirtualFileRef, Stamp]](TreeMap, 3) {
+        val f = readMapper.mapProductFile(VirtualFileRef.of(in.string()))
+        (f, readMapper.mapProductStamp(f, readStamp2(in)))
+      }
+    val sources = in.readColl[(VirtualFileRef, Stamp), TreeMap[VirtualFileRef, Stamp]](TreeMap, 3) {
+      val f = readMapper.mapSourceFile(VirtualFileRef.of(in.string()))
+      (f, readMapper.mapSourceStamp(f, readStamp2(in)))
+    }
+    val libraries =
+      in.readColl[(VirtualFileRef, Stamp), TreeMap[VirtualFileRef, Stamp]](TreeMap, 3) {
+        val f = readMapper.mapBinaryFile(VirtualFileRef.of(in.string()))
+        (f, readMapper.mapBinaryStamp(f, readStamp2(in)))
+      }
+    Stamps(products, sources, libraries)
+  }
+
+  private[this] def writeAnalyzedClass(
+      out: Serializer,
+      ac: AnalyzedClass,
+      storeApis: Boolean
+  ): Unit = {
+    out.writeBlock("analyzedClass") {
+      out.string(ac.name())
+      out.long(ac.compilationTimestamp())
+      out.int(ac.apiHash())
+      out.bool(ac.hasMacro)
+      out.string(ac.provenance())
+      val nh0 = ac.nameHashes()
+      val nh = if (nh0.length > 1 && sort) {
+        val nh = nh0.clone()
+        Arrays.sort(nh, nameHashComparator)
+        nh
+      } else nh0
+      out.writeArray("nameHashes.name", nh) { h => out.string(h.name()) }
+      out.writeArray("nameHashes.scope", nh) { h => out.byte(h.scope().ordinal().toByte) }
+      out.writeArray("nameHashes.hash", nh) { h => out.int(h.hash()) }
+      if (storeApis) {
+        val comp = ac.api()
+        writeClassLike(out, comp.classApi())
+        writeClassLike(out, comp.objectApi())
+      }
+    }
+  }
+
+  private[this] def readAnalyzedClass(in: Deserializer, storeApis: Boolean): AnalyzedClass = {
+    in.readBlock {
+      val name = in.string()
+      val ts = in.long()
+      val ah = in.int()
+      val hm = in.bool()
+      val p = in.string()
+      val nhNames = in.readStringArray()
+      val nhScopes = in.readArray[UseScope]() { UseScope.values()(in.byte()) }
+      val nhHashes = in.readArray[Int]() { in.int() }
+      val nameHashes = new Array[NameHash](nhNames.length)
+      var i = 0
+      while (i < nameHashes.length) {
+        nameHashes(i) = NameHash.of(nhNames(i), nhScopes(i), nhHashes(i))
+        i += 1
+      }
+      val comp =
+        if (storeApis) Companions.of(readClassLike(in), readClassLike(in))
+        else APIs.emptyCompanions
+      AnalyzedClass.of(ts, name, SafeLazyProxy.strict(comp), ah, nameHashes, hm, ah, p)
+    }
+  }
+
+  private[this] def writeAPIs(out: Serializer, apis: APIs, storeApis: Boolean): Unit = {
+    def write(n: String, m: Map[String, AnalyzedClass]): Unit =
+      writeMaybeSortedStringMap(out, n, m.mapValues(_.withCompilationTimestamp(-1L))) { ac =>
```

Comment:
What is the reason for setting timestamp to `-1L`?

I am not fully sure, but I think sbt `testQuick` uses analysis timestamp.

Reply:
Any metadata about a compilation run inevitably results in inconsistent outputs. We can only store data that is repeatable in a fresh build. If sbt needs this data, we could include it optionally (similar to the unsorted output that I added at the last minute).

Reply:
We should probably use 2010-01-01, like elsewhere in sbt - https://github.com/sbt/sbt/pull/6237, if not a few seconds after the midnight of 2010-01-01 (to break tie break with source JARs).
There's a weird combination ZIP + JVM bug where any timestamp before 1980 requires extended timestamp, and it sometimes end up capturing the timezone of the host machine, breaking hermeticity.